# Agency Prefix Code Profiles
Which agencies apply the most or have the most prefix codes? Which ones have the least? 

Data: 
* [Oblgiated list](https://dot.ca.gov/programs/local-assistance/reports/e-76-obligated)

* running notebook with updated locode data 'obligated_list_1202.csv'

In [1]:
import pandas as pd
from siuba import *

import numpy as np
from calitp import *

import ipywidgets as widgets
from ipywidgets import *
from IPython.display import Markdown
from IPython.core.display import display

In [2]:
df = pd.read_csv('gs://calitp-analytics-data/data-analyses/dla/e-76Obligated/obligated_list_1202.csv', low_memory=False)


/opt/conda/lib/python3.9/site-packages/google/auth/_default.py:73: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK without a quota project. You might receive a "quota exceeded" or "API not enabled" error. We recommend you rerun `gcloud auth application-default login` and make sure a quota project is added. Or you can use service accounts instead. For more information about service accounts, see https://cloud.google.com/docs/authentication/
  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)


In [3]:
df.head()

location prefix project_no             agency prepared_date  \
0  Obligated  BPMPL  5904(121)    Humboldt County    2018-12-18   
1  Obligated     ER  32D0(008)   Mendocino County    2018-12-17   
2  Obligated     ER  4820(004)    Humboldt County    2018-12-07   
3  Obligated    CML  5924(244)  Sacramento County    2018-12-11   
4  Obligated    CML  5924(214)  Sacramento County    2018-12-05   

  submit__to_hq_date hq_review_date submit_to_fhwa_date to_fmis_date  \
0         2018-12-18     2018-12-18          2018-12-18   2018-12-27   
1         2018-12-19     2018-12-20          2018-12-20   2018-12-27   
2         2018-12-21     2018-12-21          2018-12-21   2018-12-27   
3         2018-12-11     2018-12-21          2018-12-27   2018-12-27   
4         2018-12-11     2018-12-21          2018-12-27   2018-12-27   

   fed_requested  ...                                   project_location  \
0           0.00  ...                      14 Bridges In Humboldt County   
1       11508.00  ...               Comptche Ukiah Road, Cr 223 Pm 17.25   
2       45499.64  ...                                Mattole Rd Pm 43.17   
3      207002.00  ...     Fair Oaks Blvd. Between Howe Ave And Munroe St   
4           0.00  ...  Florin Rd Between Power Inn Rd. And Florin Per...   

                                        type_of_work seq  \
0        Bridge Preventive Maintenance - Deck Joints   3   
1                              Permanent Restoration   3   
2                              Permanent Restoration   5   
3  Create A Smart Growth Corridor With Barrier Se...   1   
4                                   Streetscape (tc)   3   

  date_request_initiated  date_completed_request      mpo  warning  projectID  \
0                    NaN                     NaN  NON-MPO      NaN       5904   
1             2018-12-17              2018-12-18  NON-MPO      NaN       32D0   
2             2018-12-06              2018-12-07  NON-MPO      NaN       4820   
3             2018-12-07              2018-12-07    SACOG      NaN       5924   
4             2018-11-28              2018-12-04    SACOG      NaN       5924   

   projectNO compare_id_locode  
0        121              True  
1          8             False  
2          4             False  
3        244              True  
4        214              True  

[5 rows x 30 columns]

In [4]:
#making sure the locode and agency updated in the data.
## if updated, there will only be one agency listed.
## if multiple agencies listed, go back to add_locode.ipynb and run that notebook

df>>filter(_.locode=='6000')>>count(_.agency)

agency   n
0  San Francisco Bay Area Rapid Transit District  41

## How many unique prefix codes per agency

### Top Agencies
finding the top agencies with the most unique prefix codes

In [5]:
df >> group_by(_.agency) >> summarize(n=_.prefix.nunique()) >> arrange(-_.n)>>filter(_.n>=15)

agency   n
291                                        Los Angeles  39
74                                            Caltrans  27
292                                 Los Angeles County  27
459              San Bernardino Associated Governments  24
466                                          San Diego  22
439                                   Riverside County  21
492                             San Luis Obispo County  21
569                                           Stockton  21
156                                   El Dorado County  20
337                                    Monterey County  19
114                                Contra Costa County  18
293  Los Angeles County Metropolitan Transportation...  18
382                                       Palm Springs  18
449                                         Sacramento  18
451                                  Sacramento County  18
286                                         Long Beach  17
328                                            Modesto  17
467               San Diego Association Of Governments  17
475                               San Francisco County  17
508                               Santa Barbara County  17
511                                 Santa Clara County  17
5                                       Alameda County  16
242                                        Kern County  16
549                                      Sonoma County  16
181                                             Fresno  15
182                                      Fresno County  15
221                                    Humboldt County  15
227                                    Imperial County  15
365                                            Oakland  15
438                                          Riverside  15
487                                           San Jose  15
568                                  Stanislaus County  15
584                                      Tehama County  15
599                                      Tulare County  15

In [6]:
top = df >> group_by(_.agency) >> summarize(n=_.prefix.nunique()) >> arrange(-_.n)>>filter(_.n>=20)

In [7]:
top

agency   n
291                            Los Angeles  39
74                                Caltrans  27
292                     Los Angeles County  27
459  San Bernardino Associated Governments  24
466                              San Diego  22
439                       Riverside County  21
492                 San Luis Obispo County  21
569                               Stockton  21
156                       El Dorado County  20

In [8]:
@interact
def find_agencies_top(agency=top.agency.unique().tolist()):

            count_top = (df>>filter(_.agency==agency)>> count(_.prefix)>>arrange(-_.n))
            
            display(Markdown(f"**Top Prefix Codes {agency} uses:**"))
            display(count_top)


interactive(children=(Dropdown(description='agency', options=('Los Angeles', 'Caltrans', 'Los Angeles County',…

#### Printing top 5 programs for the top agencies


In [9]:
top

agency   n
291                            Los Angeles  39
74                                Caltrans  27
292                     Los Angeles County  27
459  San Bernardino Associated Governments  24
466                              San Diego  22
439                       Riverside County  21
492                 San Luis Obispo County  21
569                               Stockton  21
156                       El Dorado County  20

In [12]:
def find_agencies_top(agency=top.agency.unique().tolist()):
    
    for agency in top.agency.unique().tolist(): 
            count_top = ((df>>filter(_.agency==agency)>> count(_.prefix)>>arrange(-_.n)).head(5)>>select(_.prefix))
            
            display(Markdown(f"**Top Prefix Codes {agency} uses:**"))
            display(count_top)
            
    return

In [13]:
agency=top.agency.unique().tolist()
find_agencies_top(agency)

**Top Prefix Codes Los Angeles uses:**

prefix
27    STP
22   HSIP
12    CML
2     ATP
5    BHLS

**Top Prefix Codes Caltrans uses:**

prefix
25  STPLN
1    CMLN
10  HP21L
23    STP
0     CML

**Top Prefix Codes Los Angeles County uses:**

prefix
13     ER
18   HSIP
1   ACSTP
4   BPMPL
23    STP

**Top Prefix Codes San Bernardino Associated Governments uses:**

prefix
21    STPLN
3      CMLN
5   CMSTPLN
12   HPLULN
0       ATP

**Top Prefix Codes San Diego uses:**

prefix
4   BPMPL
15   HSIP
8    BRLS
2    BHLS
7    BRLO

**Top Prefix Codes Riverside County uses:**

prefix
14   HSIP
3    BRLS
19  STPLN
18    STP
2    BRLO

**Top Prefix Codes San Luis Obispo County uses:**

prefix
7    BRLO
16   HSIP
5    BHLS
8    BRLS
6   BPMPL

**Top Prefix Codes Stockton uses:**

prefix
10    HSIP
6      CML
16     STP
1   ATPCML
2   ATPLNI

**Top Prefix Codes El Dorado County uses:**

prefix
5    BRLO
7     CML
14   HSIP
11     ER
18    STP

Popular codes in top agencies:
   * STP/STLN: Surface Transportation Program/Surface Transportation Program Local National Highway System
   * HSIP: Highway Safety Improvement Program 
   * CML: Congestion Mitigation & Air Quality Improvement Program 
   * BHLS: ?
   * BRLO: Highway Bridge Program (HBP) - Off Federal-aid System 

### Bottom Agencies

In [14]:
(df 
     >> group_by(_.agency) 
     >> summarize(n=_.prefix.nunique())
     >> filter(_.n<=1)
 
)

agency  n
4            Alameda Corridor Transportation Authority  1
8    Alameda County Transportation Improvement Auth...  1
9            Alameda County Waste Management Authority  1
14                                         Amador City  1
16             Amador County Transportation Commission  1
..                                                 ... ..
638                                       Westminister  1
644                                            Willits  1
645                                            Willows  1
655                                         Yountville  1
657                                               Yrts  1

[140 rows x 2 columns]

In [15]:
df >> group_by(_.agency) >> summarize(n=_.prefix.nunique()) >> arrange(_.n)>>filter(_.n<=3)

agency  n
4            Alameda Corridor Transportation Authority  1
8    Alameda County Transportation Improvement Auth...  1
9            Alameda County Waste Management Authority  1
14                                         Amador City  1
16             Amador County Transportation Commission  1
..                                                 ... ..
608                                              Ukiah  3
628                                       Walnut Creek  3
633                                        West Covina  3
639                                        Westminster  3
647                                            Winters  3

[339 rows x 2 columns]

In [16]:
bottom = df >> group_by(_.agency) >> summarize(n=_.prefix.nunique()) >> arrange(_.n)>>filter(_.n<=3)

In [17]:
bottom.sample(9)

agency  n
119       Coronado  1
87        Cerritos  2
527  Scotts Valley  1
57      Buena Park  1
73       Calistoga  1
200        Gridley  2
52          Blythe  2
595       Trinidad  1
653    Yorba Linda  2

In [18]:
agency=bottom.agency.unique().tolist()


def find_agencies_bottom(agency=bottom.agency.unique().tolist()):
    
    for agency in bottom.agency.unique().tolist(): 
            count_bottom = ((df>>filter(_.agency==agency)>> count(_.prefix)>>arrange(_.n)).head(5)>>select(_.prefix))
            
            display(Markdown(f"**Top Prefix Codes {agency} uses:**"))
            display(count_bottom)
            
    return

In [19]:
print(len(agency))

339


In [20]:
find_agencies_bottom(agency)

**Top Prefix Codes Alameda Corridor Transportation Authority uses:**

prefix
0  HPLUL

**Top Prefix Codes Alameda County Transportation Improvement Authority uses:**

prefix
0    STP

**Top Prefix Codes Alameda County Waste Management Authority uses:**

prefix
0    CML

**Top Prefix Codes Amador City uses:**

prefix
0   BRLS

**Top Prefix Codes Amador County Transportation Commission uses:**

prefix
0    CML

**Top Prefix Codes Amador Rapid Transit System uses:**

prefix
0  FTACML

**Top Prefix Codes Antelope Valley Transit Authority uses:**

prefix
0  FTACML

**Top Prefix Codes Atherton uses:**

prefix
0    STP

**Top Prefix Codes Avenal uses:**

prefix
0    HIP

**Top Prefix Codes Bishop uses:**

prefix
0  RPSTP

**Top Prefix Codes Blue Lake uses:**

prefix
0  RPSTP

**Top Prefix Codes Buena Park uses:**

prefix
0    STP

**Top Prefix Codes Calaveras Council Of Governments uses:**

prefix
0  RPSTP

**Top Prefix Codes California State University, Sacramento uses:**

prefix
0   LTAP

**Top Prefix Codes California Tahoe Conservancy uses:**

prefix
0    ATP

**Top Prefix Codes Calistoga uses:**

prefix
0   BRLO

**Top Prefix Codes Cap Soeast Con uses:**

prefix
0    STP

**Top Prefix Codes Capital Southeast Connector uses:**

prefix
0    STP

**Top Prefix Codes Carpinteria uses:**

prefix
0   BRLS

**Top Prefix Codes Chula Vista Elementary School District uses:**

prefix
0  SRTSLNI

**Top Prefix Codes Clayton uses:**

prefix
0    STP

**Top Prefix Codes Cloverdale uses:**

prefix
0    STP

**Top Prefix Codes Clovis Unified School District uses:**

prefix
0  CMLNI

**Top Prefix Codes Coachella V A G uses:**

prefix
0    ATP

**Top Prefix Codes Colma uses:**

prefix
0    CML

**Top Prefix Codes Corning uses:**

prefix
0  RPSTP

**Top Prefix Codes Coronado uses:**

prefix
0  FERPL

**Top Prefix Codes Corte Madera uses:**

prefix
0    CML

**Top Prefix Codes Council Of San Benito County Governments uses:**

prefix
0   CMLG

**Top Prefix Codes Crescent City uses:**

prefix
0     ER

**Top Prefix Codes Cypress uses:**

prefix
0    STP

**Top Prefix Codes Desert Hot Springs uses:**

prefix
0    STP

**Top Prefix Codes Diamond Bar uses:**

prefix
0  HPLUL

**Top Prefix Codes Dos Palos uses:**

prefix
0    CML

**Top Prefix Codes El Segundo uses:**

prefix
0  DEM08L

**Top Prefix Codes Emeryville uses:**

prefix
0   SRTS

**Top Prefix Codes Escalon uses:**

prefix
0    CML

**Top Prefix Codes Ferndale uses:**

prefix
0  RPSTP

**Top Prefix Codes Fillmore uses:**

prefix
0    STP

**Top Prefix Codes Foster City uses:**

prefix
0    STP

**Top Prefix Codes Fresno County Rural Transit Agency uses:**

prefix
0  CMLFTA

**Top Prefix Codes Fthll/East Tca uses:**

prefix
0  HP21L

**Top Prefix Codes Gold Coast Transit District uses:**

prefix
0  FTACML

**Top Prefix Codes Grand Terrace uses:**

prefix
0   HSIP

**Top Prefix Codes Hawaiian Garden uses:**

prefix
0   SRTS

**Top Prefix Codes Hawaiian Gardens uses:**

prefix
0    STP

**Top Prefix Codes Hawthorne uses:**

prefix
0   HSIP

**Top Prefix Codes Hermosa Beach uses:**

prefix
0    STP

**Top Prefix Codes Hillsborough uses:**

prefix
0    STP

**Top Prefix Codes Humboldt uses:**

prefix
0     ER

**Top Prefix Codes Humboldt Bay Harbor Recreation & Conservation District uses:**

prefix
0  HP21L

**Top Prefix Codes Huntington Park uses:**

prefix
0    ATP

**Top Prefix Codes Ione uses:**

prefix
0     ER

**Top Prefix Codes Jpam uses:**

prefix
0    CML

**Top Prefix Codes Kern County Superintendent Of School uses:**

prefix
0    CML

**Top Prefix Codes Kings County Area Public Transit Agency uses:**

prefix
0  FTACML

**Top Prefix Codes La Jolla Band Of Luiseno Indians uses:**

prefix
0    STP

**Top Prefix Codes Laguna Beach uses:**

prefix
0  RPSTP

**Top Prefix Codes Laguna Hills uses:**

prefix
0    STP

**Top Prefix Codes Lake Forest uses:**

prefix
0    STP

**Top Prefix Codes Lakewood uses:**

prefix
0    STP

**Top Prefix Codes Lgnw uses:**

prefix
0  RPSTP

**Top Prefix Codes Livingston uses:**

prefix
0    CML

**Top Prefix Codes Lodi Unified School District uses:**

prefix
0    CML

**Top Prefix Codes Long Beach Transportation Company uses:**

prefix
0  FTACML

**Top Prefix Codes Los Angeles Unified School District uses:**

prefix
0  ATPLNI

**Top Prefix Codes Malibu uses:**

prefix
0   HSIP

**Top Prefix Codes Marin County Transit District uses:**

prefix
0  FTASTPL

**Top Prefix Codes Marysville uses:**

prefix
0    STP

**Top Prefix Codes Maywood uses:**

prefix
0  HPLUL

**Top Prefix Codes Mendocino uses:**

prefix
0  ACSTP

**Top Prefix Codes Merced Co Aog uses:**

prefix
0   HPLU

**Top Prefix Codes Monte Sereno uses:**

prefix
0    STP

**Top Prefix Codes Morgan Hill uses:**

prefix
0    STP

**Top Prefix Codes Morongo Basin Transit Authority uses:**

prefix
0  FTACML

**Top Prefix Codes Mountain Area Regional Transit Authority uses:**

prefix
0  FTACML

**Top Prefix Codes Napa County Office Of Education uses:**

prefix
0  SRTSLNI

**Top Prefix Codes National Park Service, Pacific West Region uses:**

prefix
0   CASB

**Top Prefix Codes Natomas Unified School District uses:**

prefix
0  SRTSLNI

**Top Prefix Codes Newark uses:**

prefix
0    STP

**Top Prefix Codes Orland uses:**

prefix
0   HSIP

**Top Prefix Codes Palos Verdes Est uses:**

prefix
0   HSIP

**Top Prefix Codes Parlier Unified School District uses:**

prefix
0  CMLNI

**Top Prefix Codes Piedmont uses:**

prefix
0    STP

**Top Prefix Codes Port Hueneme uses:**

prefix
0    STP

**Top Prefix Codes Port Of Long Beach uses:**

prefix
0    CML

**Top Prefix Codes Port Of San Francisco uses:**

prefix
0    STP

**Top Prefix Codes Portola uses:**

prefix
0   SRTS

**Top Prefix Codes Poway uses:**

prefix
0    STP

**Top Prefix Codes Ranchopalosverd uses:**

prefix
0    CML

**Top Prefix Codes Red Bluff uses:**

prefix
0   BRLS

**Top Prefix Codes Rio Dell uses:**

prefix
0  RPSTP

**Top Prefix Codes Riverside Transit Agency uses:**

prefix
0  FTACML

**Top Prefix Codes Rollinghillsest uses:**

prefix
0  RPSTP

**Top Prefix Codes Samd uses:**

prefix
0  HPLUL

**Top Prefix Codes San Diego Metropolitan Tranit System uses:**

prefix
0  FTASTPL

**Top Prefix Codes San Dimas uses:**

prefix
0   BHLS

**Top Prefix Codes San Joaquin Regional Transit District uses:**

prefix
0  FTACML

**Top Prefix Codes Santa Cruz County Health Services Agency uses:**

prefix
0  SRTSLNI

**Top Prefix Codes Santa Fe Sprgs uses:**

prefix
0  DEML02

**Top Prefix Codes Santa Fe Springs uses:**

prefix
0  DEML02

**Top Prefix Codes Scotts Valley uses:**

prefix
0     ER

**Top Prefix Codes Shasta County Regional Transportation Planning Agency uses:**

prefix
0  SRTSLNI

**Top Prefix Codes Sierra Madre uses:**

prefix
0     ER

**Top Prefix Codes Sjrt uses:**

prefix
0  RPSTP

**Top Prefix Codes Solano County Transit uses:**

prefix
0  FTACML

**Top Prefix Codes Sonoma County Transit uses:**

prefix
0  FTASTPL

**Top Prefix Codes Sonora uses:**

prefix
0    CML

**Top Prefix Codes South Coast Air Quality Management District uses:**

prefix
0  HPLUL

**Top Prefix Codes South Coast Area Transit uses:**

prefix
0  FTACML

**Top Prefix Codes South El Monte uses:**

prefix
0   HSIP

**Top Prefix Codes Southern California Regional Rail Authority uses:**

prefix
0  FTACML

**Top Prefix Codes St Helena uses:**

prefix
0   HSIP

**Top Prefix Codes Stockton Port District uses:**

prefix
0   BRLS

**Top Prefix Codes Susanville uses:**

prefix
0  RPSTP

**Top Prefix Codes Sutter uses:**

prefix
0  STPLR

**Top Prefix Codes Tahoe Regional Planning Agency uses:**

prefix
0  SNPLMA

**Top Prefix Codes Tehama uses:**

prefix
0  RPSTP

**Top Prefix Codes Tehama County Transportation Commission uses:**

prefix
0  FTACML

**Top Prefix Codes Tri Delta Transit uses:**

prefix
0  FTASTP

**Top Prefix Codes Trinidad uses:**

prefix
0  RPSTP

**Top Prefix Codes Tuolumne County Transportation Council uses:**

prefix
0  FTACML

**Top Prefix Codes Tustin uses:**

prefix
0    STP

**Top Prefix Codes Twentynine Palm uses:**

prefix
0     BR

**Top Prefix Codes Union Pacific uses:**

prefix
0  TCIFESP

**Top Prefix Codes University Of California - Davis uses:**

prefix
0   BRLS

**Top Prefix Codes University Of California - Santa Barbara uses:**

prefix
0    ATP

**Top Prefix Codes University Of California At San Francisco uses:**

prefix
0  SRTSLNI

**Top Prefix Codes Usda Forest Ser uses:**

prefix
0   CASB

**Top Prefix Codes Victor Valley Transit Authority uses:**

prefix
0  FTACML

**Top Prefix Codes Villa Park uses:**

prefix
0    STP

**Top Prefix Codes Walnut uses:**

prefix
0    STP

**Top Prefix Codes Watsonville uses:**

prefix
0   HSIP

**Top Prefix Codes Weed uses:**

prefix
0  RPSTP

**Top Prefix Codes Western Shasta Resource Conservation District uses:**

prefix
0   CASB

**Top Prefix Codes Westminister uses:**

prefix
0   HSIP

**Top Prefix Codes Willits uses:**

prefix
0   BRLS

**Top Prefix Codes Willows uses:**

prefix
0  RPSTP

**Top Prefix Codes Yountville uses:**

prefix
0  RPSTP

**Top Prefix Codes Yrts uses:**

prefix
0    CML

**Top Prefix Codes Access Services uses:**

prefix
1  FTASTPLR
0   FTASTPL

**Top Prefix Codes Agoura Hills uses:**

prefix
1  HPLUN
0   BHLO

**Top Prefix Codes Alameda - Contra Costa Transit District uses:**

prefix
1  FTASTPL
0   FTACML

**Top Prefix Codes Atwater uses:**

prefix
1   HSIP
0    CML

**Top Prefix Codes Auburn uses:**

prefix
1   SRTS
0    CML

**Top Prefix Codes Bay Area Air Quality Management District uses:**

prefix
0    CML
1  CMLNI

**Top Prefix Codes Big Bear Lake uses:**

prefix
0     ATP
1  HSIPLN

**Top Prefix Codes Biggs uses:**

prefix
1    CML
0   BRLO

**Top Prefix Codes Blythe uses:**

prefix
0  SRTSLNI
1      STP

**Top Prefix Codes Brisbane uses:**

prefix
0   HSIP
1    STP

**Top Prefix Codes Burlingame uses:**

prefix
1    STP
0    CML

**Top Prefix Codes Calabasas uses:**

prefix
0     ER
1   SRTS

**Top Prefix Codes Central Unified School District uses:**

prefix
0    CML
1  CMLNI

**Top Prefix Codes Cerritos uses:**

prefix
0   HSIP
1    STP

**Top Prefix Codes Chino uses:**

prefix
0  HPLUL
1   HSIP

**Top Prefix Codes Claremont uses:**

prefix
0     HSIP
1  SRTSLNI

**Top Prefix Codes Clearlake uses:**

prefix
1   HSIP
0    ATP

**Top Prefix Codes Compton uses:**

prefix
1  HPLUL
0  HP21L

**Top Prefix Codes Cotati uses:**

prefix
1    STP
0    CML

**Top Prefix Codes Council Of Fresno County Governments uses:**

prefix
0   FTACML
1  PBICLNI

**Top Prefix Codes Cudahy uses:**

prefix
0    ATP
1   HSIP

**Top Prefix Codes Dana Point uses:**

prefix
0  RPSTP
1    STP

**Top Prefix Codes Del Mar uses:**

prefix
1  STPLZ
0   BRLS

**Top Prefix Codes Dixon uses:**

prefix
0     IM
1    STP

**Top Prefix Codes Dst Hot Springs uses:**

prefix
0   SRTS
1    STP

**Top Prefix Codes Dwr uses:**

prefix
0  BHLOZ
1  BRLOZ

**Top Prefix Codes El Cajon uses:**

prefix
1   SRTS
0   HSIP

**Top Prefix Codes Fortuna uses:**

prefix
1  RPSTP
0   HSIP

**Top Prefix Codes Fountain Valley uses:**

prefix
0   HSIP
1    STP

**Top Prefix Codes Fresno Unified School District uses:**

prefix
0    CML
1  CMLNI

**Top Prefix Codes Gridley uses:**

prefix
1  RPSTP
0    CML

**Top Prefix Codes Grover Beach uses:**

prefix
0    CML
1  RPSTP

**Top Prefix Codes Gustine uses:**

prefix
1   HSIP
0    CML

**Top Prefix Codes Gustine City uses:**

prefix
1   HSIP
0    CML

**Top Prefix Codes Half Moon Bay uses:**

prefix
1    CML
0   BRLS

**Top Prefix Codes Huron uses:**

prefix
1    STP
0    CML

**Top Prefix Codes Indian Wells uses:**

prefix
1    STP
0  BRLSN

**Top Prefix Codes Industry uses:**

prefix
1   TTDG
0  BHLSP

**Top Prefix Codes Irwindale uses:**

prefix
0     ER
1  HPLUL

**Top Prefix Codes Kern Regional Transit uses:**

prefix
0  CMLFTA
1  FTACML

**Top Prefix Codes Kings Canyon Unified School District uses:**

prefix
0    CML
1  CMLNI

**Top Prefix Codes La Canada Flintridge uses:**

prefix
1   HSIP
0    CML

**Top Prefix Codes La Habra uses:**

prefix
1    STP
0   HSIP

**Top Prefix Codes La Mirada uses:**

prefix
0   HSIP
1    STP

**Top Prefix Codes La Palma uses:**

prefix
0   SRTS
1    STP

**Top Prefix Codes La Puente uses:**

prefix
1    STP
0   HSIP

**Top Prefix Codes Laguna Niguel uses:**

prefix
0  BRLSZ
1    STP

**Top Prefix Codes Lemon Grove uses:**

prefix
1  SRTSLNI
0     SRTS

**Top Prefix Codes Live Oak uses:**

prefix
0    CML
1    STP

**Top Prefix Codes Livermore Amador Valley Transit Authority uses:**

prefix
1   RPSTP
0  FTACML

**Top Prefix Codes Loomis uses:**

prefix
1    STP
0    CML

**Top Prefix Codes Los Altos Hills uses:**

prefix
0  RPSTP
1    STP

**Top Prefix Codes Mammoth Lakes uses:**

prefix
1   SRTS
0  RPSTP

**Top Prefix Codes Marina uses:**

prefix
0   HSIP
1  RPSTP

**Top Prefix Codes Menifee uses:**

prefix
1    STP
0   HSIP

**Top Prefix Codes Millbrae uses:**

prefix
0    CML
1    STP

**Top Prefix Codes Milpitas uses:**

prefix
0    CML
1    STP

**Top Prefix Codes Mono uses:**

prefix
1  RPSTP
0  BPMPL

**Top Prefix Codes Montebello uses:**

prefix
0   HSIP
1    STP

**Top Prefix Codes Monterey uses:**

prefix
1   HSIP
0    ATP

**Top Prefix Codes Monterey Park uses:**

prefix
0   HSIP
1    STP

**Top Prefix Codes Moorpark uses:**

prefix
0    CML
1    STP

**Top Prefix Codes Morro Bay uses:**

prefix
1    CML
0   CASB

**Top Prefix Codes Needles uses:**

prefix
0  BPMPL
1   NCPD

**Top Prefix Codes Nevada City uses:**

prefix
0   BRLS
1    CML

**Top Prefix Codes Newman uses:**

prefix
1    STP
0    CML

**Top Prefix Codes Oakley uses:**

prefix
0  HPLUL
1    STP

**Top Prefix Codes Orange Cove uses:**

prefix
1    STP
0    CML

**Top Prefix Codes Paramount uses:**

prefix
1    STP
0   BRLS

**Top Prefix Codes Perris uses:**

prefix
1    STP
0   HSIP

**Top Prefix Codes Pinole uses:**

prefix
0   BRLS
1    STP

**Top Prefix Codes Plymouth uses:**

prefix
1   HRRR
0  HPLUL

**Top Prefix Codes Port Of Oakland uses:**

prefix
0     CML
1  DEM05L

**Top Prefix Codes Portola Valley uses:**

prefix
1    STP
0     ER

**Top Prefix Codes Rancho Santa Margarita uses:**

prefix
1    STP
0  BPMPL

**Top Prefix Codes Redondo Beach uses:**

prefix
0  FTACML
1     STP

**Top Prefix Codes Rsm uses:**

prefix
0  BPMPL
1   HSIP

**Top Prefix Codes Sacramento Metropolitan Air Quality Management District uses:**

prefix
1  CMLNI
0    CML

**Top Prefix Codes Sacramento Regional Transit District uses:**

prefix
1  FTASTPL
0   FTACML

**Top Prefix Codes San Bruno uses:**

prefix
1    STP
0    CML

**Top Prefix Codes San Clemente uses:**

prefix
0   HSIP
1    STP

**Top Prefix Codes San Diego Unified Port District uses:**

prefix
1  TCEPSB1L
0     FERPL

**Top Prefix Codes San Francisco Bay Area Water Transit Authority uses:**

prefix
1  FTASTPL
0   FTAFBP

**Top Prefix Codes San Joaquin Cog uses:**

prefix
0  DEM06U
1     STP

**Top Prefix Codes San Luis Obispo Regional Transit Agency uses:**

prefix
0  FTACML
1  FTAMCL

**Top Prefix Codes Sanbuenaventura uses:**

prefix
0    CML
1    STP

**Top Prefix Codes Sanjuancapistra uses:**

prefix
0   ESPL
1   SRTS

**Top Prefix Codes Santa Paula uses:**

prefix
1    STP
0    CML

**Top Prefix Codes Seaside uses:**

prefix
1  SRTSLNI
0      ATP

**Top Prefix Codes Sebastopol uses:**

prefix
0    CML
1    STP

**Top Prefix Codes Solvang uses:**

prefix
1     ER
0  BRLSZ

**Top Prefix Codes South Pasadena uses:**

prefix
1  RABA01
0    HSIP

**Top Prefix Codes Southwest Transportation Agency uses:**

prefix
0    CML
1  CMLNI

**Top Prefix Codes Stpd uses:**

prefix
0   BRLS
1   PLHL

**Top Prefix Codes Truckee uses:**

prefix
0   HPLU
1   HSIP

**Top Prefix Codes Twentynine Palms uses:**

prefix
0     BR
1   HSIP

**Top Prefix Codes University Of California uses:**

prefix
0   LTAP
1   VPPL

**Top Prefix Codes Vernon uses:**

prefix
1    HSIP
0  DBPL02

**Top Prefix Codes Vista uses:**

prefix
1  TCSPDE
0    HSIP

**Top Prefix Codes Waterford uses:**

prefix
0    CML
1    STP

**Top Prefix Codes Western Contra Costa Transit Authority uses:**

prefix
0   FTACML
1  FTASTPL

**Top Prefix Codes Westlake Village uses:**

prefix
1   HSIP
0   BRLO

**Top Prefix Codes Wheatland uses:**

prefix
1    STP
0   SRTS

**Top Prefix Codes Woodlake uses:**

prefix
0    ATP
1    CML

**Top Prefix Codes Yolo County Transportation District uses:**

prefix
1  FTASTCML
0    FTACML

**Top Prefix Codes Yorba Linda uses:**

prefix
1    STP
0   HSIP

**Top Prefix Codes Yosemite Area Regional Transportation System Jpa uses:**

prefix
1  CMLNI
0    CML

**Top Prefix Codes Yreka City uses:**

prefix
1   BRLS
0  BPMPL

**Top Prefix Codes Albany uses:**

prefix
0      CML
1  SRTSLNI
2      STP

**Top Prefix Codes Alhambra uses:**

prefix
0   HSIP
1   SRTS
2    STP

**Top Prefix Codes Amador uses:**

prefix
2   HRRR
0  BPMPL
1   BRLO

**Top Prefix Codes American Canyon uses:**

prefix
1  RPSTP
0    CML
2    STP

**Top Prefix Codes Angels uses:**

prefix
2   HSIP
1   HIPL
0    CML

**Top Prefix Codes Antioch uses:**

prefix
0   BHLS
2    STP
1    CML

**Top Prefix Codes Association Of Monterey Bay Area Governments uses:**

prefix
2  FSHRP2L
0      CML
1   FSHRP2

**Top Prefix Codes Banning uses:**

prefix
1    STP
2  STPLR
0   HSIP

**Top Prefix Codes Beaumont uses:**

prefix
0  DEM10L
2     STP
1      ER

**Top Prefix Codes Bellflower uses:**

prefix
0  HPLUL
1   HSIP
2    STP

**Top Prefix Codes Calaveras C T C uses:**

prefix
1    CML
0   CASB
2  RPSTP

**Top Prefix Codes Calexico uses:**

prefix
2    STP
1  HPLUL
0    CML

**Top Prefix Codes Central Contra Costa Transit Authority uses:**

prefix
0      FTA
1   FTASTP
2  FTASTPL

**Top Prefix Codes City/County Association Of Governments Of San Mateo County uses:**

prefix
0   CMLNI
1     STP
2  STPCML

**Top Prefix Codes Coachella Valley Association Of Governments uses:**

prefix
2  RPSTP
0    ATP
1    CML

**Top Prefix Codes Colfax uses:**

prefix
0     CML
1  HIPCML
2     STP

**Top Prefix Codes Colusa County uses:**

prefix
0  BPMPL
1     BR
2   BRLO

**Top Prefix Codes Commerce uses:**

prefix
0  DEM08L
1   HPLUL
2    HSIP

**Top Prefix Codes Corcoran uses:**

prefix
2  HIPL5
1   HIPL
0    CML

**Top Prefix Codes Covina uses:**

prefix
1   HSIP
0  HPLUL
2    STP

**Top Prefix Codes Cupertino uses:**

prefix
0  BPMPL
1   HSIP
2    STP

**Top Prefix Codes Dublin uses:**

prefix
0    CML
1  HPLUL
2    STP

**Top Prefix Codes Dunsmuir uses:**

prefix
0   BHLO
1   BRLO
2  RPSTP

**Top Prefix Codes Eastvale uses:**

prefix
2    STP
0    CML
1   HSIP

**Top Prefix Codes El Paso De Robles uses:**

prefix
2  RPSTP
0   BHLS
1    CML

**Top Prefix Codes Farmersville uses:**

prefix
0  CMHSIPL
2      STP
1      CML

**Top Prefix Codes Fort Bragg uses:**

prefix
0  ATPLNI
2    SRTS
1  DEMO6L

**Top Prefix Codes Glendora uses:**

prefix
0   BHLS
1   HSIP
2    STP

**Top Prefix Codes Golden Empire Transit District uses:**

prefix
2  FTACML14
0    CMLFTA
1    FTACML

**Top Prefix Codes Grass Valley uses:**

prefix
0  ACSTP
2  CMLNI
1    CML

**Top Prefix Codes Hanford uses:**

prefix
2   HSIP
1  CMLNI
0    CML

**Top Prefix Codes Healdsburg uses:**

prefix
2    STP
1    CML
0   BRLS

**Top Prefix Codes Hemet uses:**

prefix
1   HSIP
0   BRLS
2    STP

**Top Prefix Codes Hercules uses:**

prefix
2  TGR2DG
0     CML
1     STP

**Top Prefix Codes Hughson uses:**

prefix
2  STPCML
0     CML
1     STP

**Top Prefix Codes Imperial County Transportation Commission uses:**

prefix
2  FTACML
0  BUILDL
1     CML

**Top Prefix Codes Inyo uses:**

prefix
0   BRLO
2  RPSTP
1     ER

**Top Prefix Codes Jurupa Valley uses:**

prefix
0  ACST-ER
1       ER
2     HSIP

**Top Prefix Codes Kern County (District 9) uses:**

prefix
1  ATPLN
0    ATP
2    CML

**Top Prefix Codes La Canada-Flint uses:**

prefix
2  TCSEL08
0     BRLO
1     BRLS

**Top Prefix Codes Lathrop uses:**

prefix
0   BRLS
1    CML
2    STP

**Top Prefix Codes Lawndale uses:**

prefix
0    HPLUL
1     HSIP
2  SRTSLNI

**Top Prefix Codes Livermore uses:**

prefix
2   TCSP
0    CML
1    STP

**Top Prefix Codes Los Banos uses:**

prefix
1  HIPLPPSB
2      SRTS
0       CML

**Top Prefix Codes Los Gatos uses:**

prefix
0    CML
2    STP
1   HSIP

**Top Prefix Codes Lynwood uses:**

prefix
0    ATP
1  HPLUL
2   HSIP

**Top Prefix Codes Manhattan Beach uses:**

prefix
2   SRTS
0  HPLUL
1   HSIP

**Top Prefix Codes Manteca uses:**

prefix
1  RPSTP
0   HSIP
2    STP

**Top Prefix Codes Mono County uses:**

prefix
2   RSTP
0   CASB
1  RPSTP

**Top Prefix Codes Monrovia uses:**

prefix
1   HSIP
0     ER
2    STP

**Top Prefix Codes Montclair uses:**

prefix
0    BRLS
2  PNRSHP
1    HSIP

**Top Prefix Codes Napa County Transportation Planning Agency uses:**

prefix
0   ATPSB1
1      CML
2  FTASTPL

**Top Prefix Codes Napa Valley Transportation Authority uses:**

prefix
0     ATPLNI
2    FTASTPL
1  FTASTPCML

**Top Prefix Codes Newport Beach uses:**

prefix
2    STP
0   BRLO
1    CML

**Top Prefix Codes Novato uses:**

prefix
0   BHLS
1   HSIP
2    STP

**Top Prefix Codes Oakdale uses:**

prefix
1  CMSTPL
0     CML
2     STP

**Top Prefix Codes Ojai uses:**

prefix
0  ATPCML
1     CML
2     STP

**Top Prefix Codes Omnitrans uses:**

prefix
2  FTASTPL
0  FTAATPL
1   FTACML

**Top Prefix Codes Oroville uses:**

prefix
0  ATPCML
1     CML
2    HSIP

**Top Prefix Codes Pacifica uses:**

prefix
1  HP21L
0    CML
2    STP

**Top Prefix Codes Palm Desert uses:**

prefix
0     CML
1  HP21CM
2     STP

**Top Prefix Codes Rancho Cucamonga uses:**

prefix
1     SRTS
2  SRTSLNI
0    IMDLN

**Top Prefix Codes Ranchocucamonga uses:**

prefix
1     SRTS
2  SRTSLNI
0    IMDLN

**Top Prefix Codes Redlands uses:**

prefix
0  BPMPL
2   SRTS
1   HSIP

**Top Prefix Codes Rio Vista uses:**

prefix
1  HPLUL
2   HSIP
0    CML

**Top Prefix Codes Riverbank uses:**

prefix
2  STPCML
1     STP
0     CML

**Top Prefix Codes Rocklin uses:**

prefix
2    STP
0  BPMPL
1    CML

**Top Prefix Codes Rohnert Park uses:**

prefix
1  RPSTP
0    CML
2    STP

**Top Prefix Codes Rosemead uses:**

prefix
1   SRTS
2    STP
0   HSIP

**Top Prefix Codes San Fernando uses:**

prefix
2   TCSP
1   SRTS
0   HSIP

**Top Prefix Codes San Francisco Department Of Public Health uses:**

prefix
2  SRTSLNI
0   ATPLNI
1      CML

**Top Prefix Codes San Jacinto uses:**

prefix
0     ER
2  STPLN
1   HSIP

**Top Prefix Codes San Juan Capistrano uses:**

prefix
1   SRTS
0   ESPL
2    STP

**Top Prefix Codes San Marino uses:**

prefix
2  SRTSLNI
1     SRTS
0       ER

**Top Prefix Codes Sanger uses:**

prefix
0  BPMPL
1    CML
2    STP

**Top Prefix Codes Santa Clara uses:**

prefix
1   HSIP
2    STP
0    CML

**Top Prefix Codes Santa Cruz County Regional Transportation Commission uses:**

prefix
0  CMSTPL
1  DEM06L
2  STPLNI

**Top Prefix Codes Santee uses:**

prefix
0  BRLSZ
2   SRTS
1   HSIP

**Top Prefix Codes Sausalito uses:**

prefix
1  NMTPL
0    CML
2    STP

**Top Prefix Codes Seal Beach uses:**

prefix
0  ATPSB1L
2      STP
1      CML

**Top Prefix Codes Shasta Lake uses:**

prefix
1   BRLS
0  BPMPL
2   HSIP

**Top Prefix Codes Sonoma County Transportation Authority uses:**

prefix
0    CML
2  HPLUL
1  CMLNI

**Top Prefix Codes Stan Area Aog uses:**

prefix
0    CML
1  HPLUL
2    STP

**Top Prefix Codes Sutter Creek uses:**

prefix
2   HSIP
1   BRLS
0   BRLO

**Top Prefix Codes Transit Joint Powers Authority For Merced County uses:**

prefix
1   CMLNI
2  FTACML
0     CML

**Top Prefix Codes U.S. Forest Service, Pacific Southwest Region uses:**

prefix
1       FTA
2  USFSSTPL
0      CASB

**Top Prefix Codes Ukiah uses:**

prefix
2  RPSTPLE
1    RPSTP
0     HSIP

**Top Prefix Codes Walnut Creek uses:**

prefix
0   BHLS
1   BRLS
2    STP

**Top Prefix Codes West Covina uses:**

prefix
0  HPLUL
1   HSIP
2    STP

**Top Prefix Codes Westminster uses:**

prefix
2    STP
0    ATP
1   HSIP

**Top Prefix Codes Winters uses:**

prefix
1  HPLUL
2    STP
0   BRLS

Popular codes in bottom agencies: 
   * FTACML
   * STP/RPSTP
   * HSIP
   * CML 
   * BRLS
   * BRLO
   * CMLNI
    

* 140 agencies with 1 unique prefix code.
* 339 agencies with 3 or less prefix codes. 

## In which Caltrans Districts do these agencies have the most obligations?

### For the agencies in districts that applied the least

* joining the bottom agencies to the full df
* grouping by dsitrict to count the number of obligation occurances 

In [37]:
bottom = (df 
     >> group_by(_.agency) 
     >> summarize(n=_.prefix.nunique())
     >> filter(_.n<=3)
)

In [38]:
bottom_join = inner_join(df, bottom, on = "agency")

In [39]:
bottom_dist = bottom_join >> group_by(_.dist) >> count(_.dist) >> arrange(-_.n)

In [40]:
bottom_dist.rename(columns={'n': 'bottom_n'}, inplace=True)

In [41]:
bottom_dist

dist  bottom_n
6      7       409
3      4       392
9     10       376
5      6       264
2      3       261
7      8       209
10    11       129
11    12       114
4      5        86
8      9        74
0      1        61
1      2        52

### For the agencies in districts that applied the most

* joining the top agencies to the full df
* grouping by dsitrict to count the number of obligation occurances 

In [42]:
top_join = inner_join(df, top, on = "agency")

In [43]:
top_dist = top_join >> group_by(_.dist) >> count(_.dist) >> arrange(-_.n)

In [44]:
top_dist.rename(columns={'n': 'top_n'}, inplace=True)

In [45]:
top_dist

dist  top_n
7      7    960
3      3    312
10    10    262
8      8    173
5      5    133
11    11    125
12    12     13
4      4     10
0      0      4
6      6      3
1      1      2
2      2      1
9      9      1

### Number of Obligations for Top and Bottom on District
 
* joining the dataframes we just made to compare the top and bottom easier
* finding which districts have the most obligations 

In [46]:
dist = full_join(bottom_dist, top_dist, on = "dist")

In [47]:
dist>>arrange(_.dist)

dist  bottom_n  top_n
12     0       NaN      4
10     1      61.0      2
11     2      52.0      1
4      3     261.0    312
1      4     392.0     10
8      5      86.0    133
3      6     264.0      3
0      7     409.0    960
5      8     209.0    173
9      9      74.0      1
2     10     376.0    262
6     11     129.0    125
7     12     114.0     13

In [48]:
dist.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13 entries, 0 to 12
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   dist      13 non-null     int64  
 1   bottom_n  12 non-null     float64
 2   top_n     13 non-null     int64  
dtypes: float64(1), int64(2)
memory usage: 416.0 bytes


In [49]:
dist['bottom_n'] = dist['bottom_n'].fillna(0)

In [50]:
dist['bottom_n'] = dist['bottom_n'].astype(int)

In [51]:
dist.rename(columns={'dist': 'district'}, inplace=True)

In [52]:
dist>>arrange(_.district)

district  bottom_n  top_n
12         0         0      4
10         1        61      2
11         2        52      1
4          3       261    312
1          4       392     10
8          5        86    133
3          6       264      3
0          7       409    960
5          8       209    173
9          9        74      1
2         10       376    262
6         11       129    125
7         12       114     13

## Findings:



* **339 agencies** have **3 or less** unique prefixes. 
* **80 agencies** have **10 or more** unique prefixes, with the following **9 agenceies** listed with **20 or more**. 
* **District 7 and District 2** have the most obligations out of the top and bottom agencies 



| Top Agencies  | Bottom Agencies (sample of 9)|
| ------------- | ------------- | 
| Los Angeles | Sonoma County Transportation Authority | 
| Caltrans  |Port Of Long Beach |   
| Los Angeles County |Newport Beach |    
| San Bernardino Associated Governments | Portola | 
| San Diego | Calaveras Council Of Governments | 
| Riverside County | Hawthorne | 
| San Luis Obispo County | Corcoran |  
| Stockton | Claremont | 
| El Dorado County | El Cajon |  


